### PR TITLE
feat(bzlmod): support cross-platform whl setups within the main hub repo

### LINKED ISSUE #1876: Consider supporting armv7-unknown-linux-gnueabi/hf toolchain
# Feature request

It seems that the latest [standalone python release][indygreg] contains a
toolchain for armv7 for Python 3.11 and Python 3.12. In #1770 a new platform
definition has been added and it seems that the platform labels that are used
in the release are different. We have `armv7-unknown-linux-gnu` and the
published toolchain binaries are for:
* `armv7-unknown-linux-gnueabi`
* `armv7-unknown-linux-gnueabihf`

I am not sure what would be the right thing to do here yet as I am not using
`armv7` toolchains yet, but maybe @UebelAndre could help us out here with providing
requirements (PRs are also welcome) on how we could use the upstream toolchains.

[indygreg]: https://github.com/indygreg/python-build-standalone/releases/tag/20240224

### PR DESCRIPTION
**Pull request by aignas**
With this change we add support for platform-specific wheel registration and
doing the selection of which wheel is used at build time.

This supports:
- Different package versions for different platforms.
- Use string_flags to configure what to fetch/select:
    - only whls, only sdist or auto mode.
    - libc version and `musl` vs `glibc` selection.
    - universal2 vs arch wheels for mac.
    - target osx version selection.

Summary of changes:
- The `uv pip install` would only warn the user of yanked packages but
  would not refuse to install them. Update our implementation to better
  match the same behaviour.
- A previous PR has added the support for passing it in the
  `requirements_by_platform` and this just add the necessary code to make
  sure that we can also do the dependency management when parsing the
  `whl` `METADATA` files.
- Only configure `dev_pip` deps for `linux` and `osx` platforms to not raise
  issues later.
- Add a function for generating a `whl_library` name from a `filename`.
- Add a macro for generating all config settings for a particular set of parameters.
- Update `render_pkg_aliases` to also use those config settings.
- Update the toolchain selection `target_settings` to use the `py_linux_libc`
  config setting. With this the user can register a `musl` linux toolchain if
  needed. We can also use similar `flag_values` to resolve #1876.
- Integrate everything in the `pip` extension and setup cross-platform
  select statements.

Work towards #1357, #260

Stacked on #1937

### CODE REVIEW

**Comment by aignas** [comment]
I think I've identified at least multiple changes that can be separate PRs:
* refactor groups to be created inside the hub repo
* add tests for the index downloader and add a flag to control wether the downloading is happening in parallel. I've noticed it to be much easier to debug.
* use skylib for selects in the alias repo.
*Add multi platform requirement support with all of the available wheels registered.

**Comment by aignas** [comment]
This now needs to be reimplemented/rebased on #1875 and #1885.

**Comment by aignas** [comment]
@dougthor42, if you have time, I'd be curious if you could try this PR, It may have the same issue as #1917, but if does not, then it would be a really interesting data point.

**Comment by dougthor42** [comment]
A different error than previously seen. This one is:

```
ERROR: Traceback (most recent call last):
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/bzlmod/pip.bzl", line 481, column 30, in _pip_impl
                _create_whl_repos(module_ctx, pip_attr, hub_whl_map, whl_overrides, hub_group_map, simpleapi_cache)
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/bzlmod/pip.bzl", line 276, column 32, in _create_whl_repos
                whl_library(name = repo_name, **dict(sorted(whl_library_args.items())))
Error in repository_rule: invalid user-provided repo name 'pypi_311__REDACTED_0_1_0+7_g9c9466d_py3_none_any': valid names may contain only A-Z, a-z, 0-9, '-', '_', '.', and must start with a letter
ERROR: Analysis of target '//src/pyle/dataking/system_optimization/snake_optimizer:snake_metrics_test' failed; build aborted: error evaluating module extension pip in @@rules_python~//python/extensions:pip.bzl
```

Some of our internal packages include the short git commit hash in their wheel name.

**Comment by aignas** [comment]
This still needs an updated git commit message and more docs, but the functionality is there and is covered by tests, so it should be ready to test/review.

I'll add more docs later.

Sorry for the big PR, but I am not sure if it is possible to split it without loosing the context that is happening all over the `pip.parse` extension.

This should be portable to the `WORKSPACE` world, but I would like to leave this out of scope for now.

**Comment by aignas** [comment]
Right now the only tests that fail are on Windows and they fail with a weird:
```
(15:03:02) WARNING: C:/b/bk-windows-nznf/bazel/rules-python-python/docs/BUILD.bazel:47:6: target '//docs:requirements_parser_bzl' is deprecated: Use //python/pip_install:pip_repository_bzl instead; Both the requirements parser and targets under //docs are internal
(15:03:14) ERROR: C:/b/bk-windows-nznf/bazel/rules-python-python/tests/config_settings/transition/BUILD.bazel:6:25: in _transition_py_binary rule //tests/config_settings/transition:test_py_binary_with_transition_subject:
Traceback (most recent call last):
	File "C:/b/bk-windows-nznf/bazel/rules-python-python/python/config_settings/transition.bzl", line 78, column 13, in _transition_py_impl
		fail("target {} does not have rules_python PyRuntimeInfo or builtin PyRuntimeInfo".format(target))
Error in fail: target <target //tests/config_settings/transition:_test_py_binary_with_transition_subject> does not have rules_python PyRuntimeInfo or builtin PyRuntimeInfo
(15:03:14) ERROR: C:/b/bk-windows-nznf/bazel/rules-python-python/tests/config_settings/transition/BUILD.bazel:6:25: in _transition_py_test rule //tests/config_settings/transition:test_py_test_with_transition_subject:
Traceback (most recent call last):
	File "C:/b/bk-windows-nznf/bazel/rules-python-python/python/config_settings/transition.bzl", line 78, column 13, in _transition_py_impl
		fail("target {} does not have rules_python PyRuntimeInfo or builtin PyRuntimeInfo".format(target))
```

I'll finish here for the day. I might split out the last commit out as a separate PR since that is a small distinct change that is related to the big multi-platform change, but we can land it separately.

**Comment by aignas** [comment]
Done, split out the lock-file related work into #1937.

**Comment by aignas** [comment]
Oh, I see I'll need to resolve conflicts, but I hope to do that once I get a review and once #1937 is merged. I have reworked the commit history to be in logical chunks, so if you would like me to split this PR into multiple based on the commits, let me know.

**Comment by aignas** [comment]
Merged the main since I saw some `auto-detecting` toolchain changelog present, so just to make sure that I am not missing out on anything.

File: tests/private/pip_repo_name/pip_repo_name_tests.bzl
Comment (aignas) [inline tests/private/pip_repo_name/pip_repo_name_tests.bzl:24]: Using sha is good enough for the first pass, but PR welcome for including the version (and optionally the build tag) in the repo name.

File: python/private/pip_config_settings.bzl
Comment (aignas) [inline python/private/pip_config_settings.bzl:76]: An idea from rules_pycross: https://github.com/jvolkman/rules_pycross/blob/531c8d21ff859e8b6df7f267eef066150723ec04/pycross/private/interpreter_version.bzl

File: python/private/toolchains_repo.bzl
Comment (aignas) [inline python/private/toolchains_repo.bzl:73]: This is breaking the Windows build due to some reason, so we may want to revert this commit and split it into a separate PR.

Comment (aignas) [reply python/private/toolchains_repo.bzl:73]: Found the issue and fixed in https://github.com/bazelbuild/rules_python/pull/1837/commits/8a6bda59e95466c4922ca7c32392200d5c19c008

File: python/private/pip_config_settings.bzl
Comment (aignas) [reply python/private/pip_config_settings.bzl:76]: This could be a good alternative to the current `is_python_config_setting` implementation as it would reduce the number of flags that we generate.

File: tests/config_settings/transition/multi_version_tests.bzl
Comment (aignas) [inline tests/config_settings/transition/multi_version_tests.bzl:26]: The tests were working previously because even though the test is setting a specific `python_version` via a transition, the toolchain that we were getting was the default version. So in this case the actual version would be returned would be different from the one requested in the transition.

This could explain some of the issues we've seen recently about people saying that the toolchain is mismatching their configuration as it could happen if they do not configure the default toolchain version in their `MODULE.bazel` file.

I wonder if this should be mentioned in the `CHANGELOG.md`. I'll do it for the sake of posterity.

**Comment by dougthor42** [comment]
Still failing on the `foobar` package that has a md5 for both py311 and py312 :face_with_diagonal_mouth:

It's a different failure, though, so that's good!

```console
$ bazel clean; bazel build //...
INFO: Invocation ID: cefb7077-a468-4283-b711-06897aff5e1d
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
INFO: Invocation ID: 84fe841f-1ab2-4990-9261-165044c971ac
ERROR: /usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~~pip~pypi/_config/BUILD.bazel: no such target '@@rules_python~~pip~pypi//_config:is_py311_none_manylinux_x86_64': target 'is_py311_none_manylinux_x86_64' not declared in package '_config' defined by /usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~~pip~pypi/_config/BUILD.bazel (did you mean is_py3_none_manylinux_x86_64, _is_py3_none_manylinux_x86_64, is_py_none_manylinux_x86_64, _is_py_none_manylinux_x86_64, is_cp3x_none_manylinux_x86_64, or is_py3_none_any_linux_x86_64?)
ERROR: /usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~~pip~pypi/foobar/BUILD.bazel:53:6: errors encountered resolving select() keys for @@rules_python~~pip~pypi//foobar:whl
Use --verbose_failures to see the command lines of failed build steps.
ERROR: Analysis of target '//:modules_map' failed; build aborted: Analysis failed
INFO: Elapsed time: 1.091s, Critical Path: 0.63s
INFO: 18 processes: 18 internal.
ERROR: Build did NOT complete successfully
```

**Comment by aignas** [comment]
@dougthor42, sorry for that one, now should be fixed

**Comment by dougthor42** [comment]
Yup! 445337e worked for me :+1: 

I noticed that the runfiles directory now has more info in the path name:
+ the wheel name
+ the head of the wheel sha

```diff
-test.runfiles/rules_python~~pip~pypi_311_foobar/
+test.runfiles/rules_python~~pip~pypi_311_foobar_py311_none_manylinux_2_17_x86_64_b7d2f441/
```

These are great changes! Super useful. Should such a change be included in the changelog?

File: tests/private/pip_repo_name/pip_repo_name_tests.bzl
Comment (groodt) [reply tests/private/pip_repo_name/pip_repo_name_tests.bzl:24]: We will need to keep this experimental until we settle on the names. Or at least remember to update if we change it and it breaks for users.

**Review by groodt — APPROVED** [review]
LGTM

As mentioned at the maintainers review, we might need to keep exploring alternatives for the explosion of config flags, but I don't think it needs to be blocking while we iterate.

**Comment by aignas** [comment]
@groodt, thanks for the stamp, could you also stamp #1937 please?

File: python/private/py_toolchain_suite.bzl
Comment (aignas) [inline python/private/py_toolchain_suite.bzl:74]: The absence of this branch is the root cause of #1967.

**Pull request merged**
